### PR TITLE
Update logic to factor in received dungeon and boss rewards

### PR DESCRIPTION
**Pull request by MattEqualsCoder**
This basically changes Rewards to be treated similarly to Items so that they can be tracked in the Progression and be factored in for logic. In order for this logic to work with multiworld in the future, this actually involved having to make a world based reward class, so I changed the old Reward enum to RewardType and created a new Reward class for this purpose.

Because unlike the previous PR, the actual logic is changed this time, I don't believe this PR is quite ready. Currently the tracker will no longer show things like Wrecked ship as available until you beat Phantoon. I'm not sure of the best solution for this currently. I'm thinking either I could add a boss/reward indicator on the map that shows up when it's in logic or I may need to add another delegate like canAccess to locations that uses the old logic to show that it's accessible but only after beating the prerequisite boss.

Along with this, still need to do some clean up and commenting. Just putting this out there to see what you think of the general implementation.

### CODE REVIEW

**Review by Vivelin — DISMISSED** [review]
This looks a lot cleaner to me. Thanks for humoring me.

**Comment by MattEqualsCoder** [comment]
No problem. In regards to the loss of functionality, my current thought is to introduce another delegate similar to canAccess. This would be something like "canAccessAfterBosses" or something along those lines. If a location is not cleared and not accessible, this would be fallback logic. I would then use that to do something like display those locations slightly differently. I'm thinking maybe just have it appear like a normal location but set the opacity to 75% or something.

This is similar to how I think the viewable locations would work.

**Comment by Vivelin** [comment]
Hmm. Maybe simply calling it `isRelevant` or `isRelevantWhen` or something. A way to determine when something is relevant and should be shown on the map regardless of logic.

We could reuse that mechanism if there are any dungeons with other complicated logic. A lot of it is covered by assuming keys, but wasn't there something with Ice Palace, too?

**Comment by MattEqualsCoder** [comment]
Good point on making it more generic in case we think of other scenarios where this could be useful to have.

In regards to ice palace, the only thing I can think of that you may be referring to is that it can require the hookshot, but it's all based on where the keys are located. That can all be bypassed with ice breaker (and maybe the ice palace bomb jump?). AFAIK, this is one of the few locations where a boss can be accessible when not all locations in the dungeon are accessible. Turtle Rock is the only other one I can think of.

Long story short, I don't know if that _isRelevant_ here as I think Ice Palace should show up as orange there currently, which I think is the most appropriate behavior as it would be true for someone new who doesn't know the tech.

**Comment by MattEqualsCoder** [comment]
Okay, with this I think the bulk of the changes are done. Ended up being a lot more extensive than I expected... Will probably be best to put this out there as a beta or release candidate first as I did a good amount of shuffling around of the logic to allow for the relevant checks.

**Comment by Vivelin** [comment]
I mean, with my refactoring and the Shaktool changes, who knows what could've broken. I might have some other big changes I want to get out of the way, like upgrading to .NET 6 and using nullable reference types everything consistently. It might be worth waiting for that a little bit.

File: src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs [hunk 1]
```diff
@@ -141,17 +159,52 @@ public ImageSource IconImage
...
-                if (ClearableLocationsCount > 0 && ClearableLocationsCount == UnclearedLocationsCount)
-                {
-                    image = "accessible.png";
-                }
-                else if (ClearableLocationsCount > 0 && ClearableLocationsCount < UnclearedLocationsCount)
+
+                if (Locations == null)
                 {
-                    image = "partial.png";
+                    var region = (IHasReward)Region;
```
Comment (Vivelin) [inline src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs:165]: I can't tell from the diff, but can't this fail? Or is `Region` always a dungeon or SM boss area here?

File: src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs [hunk 2]
```diff
@@ -199,7 +252,24 @@ public ImageSource NumberImage
         /// Get the tag to use for the location. Use the region for dungeons
         ///  and the list of locations for all other places
         /// </summary>
-        public object Tag => Region.Name == Name ? Region : Locations.Where(x => Syncer.IsLocationClearable(x, true, Syncer.Tracker.World.Config.Keysanity)).ToList();
+        public object Tag
+        {
+            get
+            {
```
Comment (Vivelin) [inline src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs:258]: I think this could be more readable as a switch expression

File: src/Randomizer.Shared/Enums/LocationStatus.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Randomizer.Shared.Enums
+{
+    /// <summary>
+    /// Indicates the current tracking status of the location
+    /// </summary>
+    public enum LocationStatus
```
Comment (Vivelin) [inline src/Randomizer.Shared/Enums/LocationStatus.cs:12]: This enum should probably have a reasonable "zero" value. "Cleared" being the default seems a little... wrong, somehow.

File: src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs (hunk 1 above)
Comment (MattEqualsCoder) [reply src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs:165]: It would only fail in the case of a bad config, and I doubt anyone would be modifying the maps config file.

File: src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs (hunk 2 above)
Comment (MattEqualsCoder) [reply src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs:258]: I don't have a common type here to switch off of, unless I'm misunderstanding you (or am unfamiliar with something fancy in C#)

File: src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs (hunk 1 above)
Comment (Vivelin) [reply src/Randomizer.App/ViewModels/TrackerMapLocationViewModel.cs:165]: So it could still conceivably come up, then. I do want to throw a better error, but for now we could probably just leave this as TODO.

**Review by Vivelin — DISMISSED** [review]

**Review by Vivelin — APPROVED** [review]
